### ISSUE TITLE
Add a  `--wait` flag as global command

### ISSUE THREAD

**Issue by whyrusleeping** [ipfs/go-ipfs#3114]
It was discussed in irc that it would be very nice to be able to start a command but not wait for it to return.  If we implement `ipfs diag cmds cancel <job>` and `ipfs diag cmds wait <job>` we would have all the tools we need to deal with 'backgrounded' commands.

cc @pjz

**Comment by Kubuxu** [comment]
We would also have to return job number somehow.

**Comment by skorokithakis** [comment]
This (or rather, not this, but what it implies) is crucial for us in the development of [Eternum](https://www.eternum.io/) (an IPFS pinning service). Our problem is that, currently, pinning things launches a task that may just time out if the file is too long, so there's no way to reliably charge for pinning. if we charge after the file has been downloaded, we might spend a month downloading a huge file without charging anything. If we charge before, we might be charging for a transfer that will time out or be interrupted at some point.

Also, if the user unpins the file before it's finished pinning, the pinning doesn't get cancelled. A job system for pinning files would solve all our problems, as then they would become job management problems instead of having to manage everything ourselves.

Basically, if pinning worked like a torrent client does, with starting/pausing/stopping/deleting, it would be ideal for us. As it stands, we have a very large problem even delivering on our promised services.

**Comment by forstmeier** [comment]
Hi, all! I saw that this related to the feature requested in #4041 but I wasn't sure of it's status - are these commands currently being worked on or does it still need to be built?

**Comment by whyrusleeping** [comment]
@forstmeier this still needs to be built. There is a small amount of design still yet to be done around the UX and API of this, but it should be pretty straightforward, all the hard bits and pieces are pretty much there already.

The main questions are "what do we name the flag?"  and "how do we return the job ID?"

**Comment by forstmeier** [comment]
@whyrusleeping Cool! So I understand (and correct me where I'm wrong):
- The overall goal here is to allow commands to be non-blocking and a global flag will be added to allow that to happen. In order to support that, additional commands are needed to control those processes moved to the background (listed below).
 1. returning running command/job IDs (like the Linux `jobs` command)
 2. cancelling running background jobs by their job ID (#3115)
 3. a `wait` command (mentioned in #3116 but I need some clarification on it)

Let me know if I'm missing something in that list.

> The main questions are "what do we name the flag?" and "how do we return the job ID?"

To address your questions while thinking out loud:
- Linux has the `&` option which allow the terminal to run the command in the background but for more verbosity maybe `--async`? I didn't see anything like that already when running `ipfs commands` but maybe it's not preferred.
- The `ipfs diag cmds -v` returns all in/active commands that have been run on that daemon - maybe adding another flag (`-a, --active`) just returns the ones running and the ID returned with `-v` can be used for cancelling/etc.? Just thoughts and I'd be happy to start poking at this if that sounds good. :wrench:
